### PR TITLE
Correct optimum start for partial hour fan start operation

### LINKED ISSUE #8013: Optimum start starting too early when fan start time is not on the hour
Issue overview
--------------
When the fan schedule is scheduled to begin on the half hour, the optimum start allows the fans to begin running a half hour earlier than expected. For example, if the optimum start is set for 1.5 hours early allowable operation and the fan schedule starts at 5:30am, the optimum start will actually allow the fans to start at 3:30am rather than the expected 4:00am. There is also a 30 minute lag between the optimum start and when the fan schedule begins where the fans actually shut off. It seems that the optimum start does not have the capability to run off partial hour schedules, and will therefore begin the optimum start from the previous whole hour.
 
image
image
image

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus v9.3

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by matthew-larson**
Pull request overview
---------------------
 - Fixes #8013 
 - This pull request modifies the FanStartTime calculation to account for partial hour fan starts in the fan operation schedule.  The `(1 / NumOfTimeStepInHour * J)` was being truncated as an integer, resulting in the FanStartTime only being able to output on the hour.  I also added the `- 0.01` at the end to account for small decimal place differences.  I added an additional check in the optimum start unit test that uses an additional zone and system I created that has a partial hour fan start time to ensure the optimum start is available during the correct times.  Below shows the objects used for my test model including the AvailabilityManager:OptimumStart object and the associated fan operation schedule with the partial hour fan start at 5:30am.  Below that shows the before and after result of the pull request, with the latter showing the correct operation with a fan start time of 5:30am and a maximum optimum start time of 1.5 hours, which results in the fan turning on at 4:00am.

Optimum Start Control:
image
image

Before:
image

After:
image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally

### CODE REVIEW

**Comment by matthew-larson** [comment]
The diffs in OptimalStart_RefBldgLargeOfficeNew2004_Chicago are expected due to the change in the `FanStartTime` calculation.  Previously it could only result in an integer which didn't correctly deal with multiple timesteps, especially ones that resulted in non-terminating decimals because of the <= comparison when deciding whether the `CurrentTime` was within the optimum start period.  Since the warmup period includes optimum start, the zone temperatures are effected starting on the first timestep.

File: src/EnergyPlus/SystemAvailabilityManager.cc [hunk 1]
```diff
@@ -2738,21 +2736,21 @@ namespace SystemAvailabilityManager {
             FanStartTime = 0.0;
             FanStartTimeTmr = 0.0;
             exitLoop = false;
-            for (I = 1; I <= 24; ++I) {
-                for (J = 1; J <= NumOfTimeStepInHour; ++J) {
+            for (int I = 1; I <= 24; ++I) {
```
Comment (mitchute) [inline src/EnergyPlus/SystemAvailabilityManager.cc:2739]: Proper variable scoping is always a good sign.

File: src/EnergyPlus/SystemAvailabilityManager.cc [hunk 2]
```diff
@@ -2738,21 +2736,21 @@ namespace SystemAvailabilityManager {
...
+                for (int J = 1; J <= NumOfTimeStepInHour; ++J) {
                     if (DayValues(J, I) <= 0.0) continue;
-                    FanStartTime = I - 1 + 1 / NumOfTimeStepInHour * J;
+                    FanStartTime = I - 1 + 1.0 / NumOfTimeStepInHour * J - 0.01;
```
Comment (mitchute) [inline src/EnergyPlus/SystemAvailabilityManager.cc:2742]: Pretty straightforward.

File: tst/EnergyPlus/unit/SystemAvailabilityManager.unit.cc [hunk 3]
```diff
@@ -320,7 +362,15 @@ TEST_F(EnergyPlusFixture, SysAvailManager_OptimumStart)
...
-    // Check that the system restores setpoints to unoccupied setpoints and don't use occupied setpoints post-occupancy
+    // #8013 - Check that the optimum start is available during the correct times when using a partial hour fan start
+    DataGlobals::CurrentTime = 5.00;   // set the current time to 5:00 AM, before max optimum start time
+    SystemAvailabilityManager::ManageSystemAvailability();
+    EXPECT_FALSE(DataHVACGlobals::OptStartData.OptStartFlag(6)); // avail manager should be set to no action for Zone 6
+    DataGlobals::CurrentTime = 6.50;   // set the current time to 6:30 AM when occupancy begins
+    SystemAvailabilityManager::ManageSystemAvailability();
+    EXPECT_TRUE(DataHVACGlobals::OptStartData.OptStartFlag(6)); // avail manager should be set to cycle on for Zone 6
+
+    // #7866 - Check that the system restores setpoints to unoccupied setpoints and don't use occupied setpoints post-occupancy
```
Comment (mitchute) [inline tst/EnergyPlus/unit/SystemAvailabilityManager.unit.cc]: Good test.

**Review by mitchute — APPROVED** [review]
@matthew-larson this looks good. Pull develop in again and we'll get this merged.

**Comment by mitchute** [comment]
Everything looks good here. Diffs on OptimalStart_RefBldgLargeOfficeNew2004_Chicago are expected. Merging.

**Pull request merged**